File: leetcode/questions/1101/python/1101.py
from typing import List

class DisjointSet:
    def __init__(self, num_groups):
        self._num_groups = num_groups
        self._groups = [group for group in range(num_groups)]
        self._ranks = [0 for _ in range(num_groups)]

    def find(self, a):
        if self._groups[a] == a:
            return a
        self._groups[a] = self.find(self._groups[a])

        return self._groups[a]

    def is_same(self, a, b):
        return self.find(a) == self.find(b)

    def unite(self, a, b):
        if self.is_same(a, b):
            return self._num_groups

        self._num_groups -= 1
        group_a = self.find(a)
        group_b = self.find(b)

        if self._ranks[group_a] > self._ranks[group_b]:
            self._groups[group_b] = group_a
        elif self._ranks[group_a] < self._ranks[group_b]:
            self._groups[group_a] = group_b
        else:
            self._groups[group_b] = group_a
            self._ranks[group_a] += 1

        return self._num_groups

    def num_groups(self):
        return self._num_groups


class Solution:
    def earliestAcq(self, logs: List[List[int]], n: int) -> int:
        disjoint_set = DisjointSet(n)
        logs.sort()

        for ts, a, b in logs:
            disjoint_set.unite(a, b)
            if disjoint_set.num_groups() == 1:
                return ts

        return -1

File: leetcode/questions/1101/java/1101.java
import java.util.*;

class DisjointSet {
    private int[] groups;
    private int[] ranks;
    private int numGroups;

    public DisjointSet(int numGroups) {
        this.numGroups = numGroups;
        this.groups = new int[numGroups];
        this.ranks = new int[numGroups];

        for (int i = 0; i < numGroups; ++i) {
            this.groups[i] = i;
            this.ranks[i] = 0;
        }
    }

    public int find(int val) {
        if (this.groups[val] == val) {
            return val;
        }

        this.groups[val] = find(this.groups[val]);
        return this.groups[val];
    }

    public boolean same(int valA, int valB) {
        return find(valA) == find(valB);
    }

    public int unite(int valA, int valB) {
        if (same(valA, valB)) {
            return this.numGroups;
        }

        int groupA = find(valA);
        int groupB = find(valB);
        this.numGroups -= 1;

        if (this.ranks[groupA] > this.ranks[groupB]) {
            this.groups[groupB] = groupA;
        } else if (this.ranks[groupA] < this.ranks[groupB]) {
            this.groups[groupA] = groupB;
        } else {
            this.groups[groupB] = groupA;
            this.ranks[groupA] += 1;
        }

        return this.numGroups;
    }
}

class Solution {
    public int earliestAcq(int[][] logs, int n) {
        DisjointSet disjointSet = new DisjointSet(n);
        PriorityQueue<int[]> priorityQueue = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        for (int[] log: logs) {
            priorityQueue.offer(log);
        }

        while(!priorityQueue.isEmpty()) {
            int[] log = priorityQueue.poll();
            int ts = log[0], a = log[1], b = log[2];
            int remain = disjointSet.unite(a, b);

            if (remain == 1) {
                return ts;
            }
        }

        return -1;
    }
}